file_store.go

```diff
@@ -0,0 +1,208 @@
+package raft
+
+import (
+	"encoding/gob"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// FileStore implements the LogStore and StableStore interface.
+type FileStore struct {
+	l         sync.RWMutex
+	lowIndex  uint64
+	highIndex uint64
+	dir       string
+}
+
+type uint64Slice []uint64
+
+func (p uint64Slice) Len() int           { return len(p) }
+func (p uint64Slice) Less(i, j int) bool { return p[i] < p[j] }
+func (p uint64Slice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func NewFileStore(dir string) (*FileStore, error) {
+	if err := os.MkdirAll(filepath.Join(dir, "logs"), 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Join(dir, "stable"), 0700); err != nil {
+		return nil, err
+	}
+
+	return &FileStore{
+		dir: dir,
```

Comment:
If all the nodes shutdown and restart (e.g. power loss to DC), then all of them will assume they have no data (high index is zero). This would be a "data loss" scenario which is easily avoidable. The indexes must be restored to prevent data loss and electing the correct leader with the most recent data.